### PR TITLE
feat(snake/core): pure logic core (config + RNG + state + tests)

### LINKED ISSUE #15: snake: pure logic core
### Problem
We need the second classic game after Tetris. Snake is the lowest-risk candidate (grid logic, no animation timing, deterministic). Per the project rule, the rules layer must be a pure GDScript module with zero `Node`/OS dependencies, fully unit-testable.

### Goal
A pure `SnakeGameState` module that fully captures Snake mechanics and is driven only by tick + intent input. No rendering, no Godot scene types.

### Scope
- Configurable grid (default 20×20), wrap or wall mode (configurable; default = walls = death).
- Snake represented as an ordered deque of cells; head at index 0.
- Direction queue: at most one buffered turn between two ticks (prevents 180° instant reverse).
- Food spawning: deterministic from a seeded RNG; never spawns inside the snake.
- Tick model: caller advances state by `tick(now_ms)`, with a configurable `step_ms` that decreases as score grows (level curve).
- Scoring: +1 per food eaten; level increases every N foods (default 5), step_ms shrinks by a factor.
- Game over: head hits wall (in walls mode) or hits its own body.
- Pause is *not* a state-machine concern — caller simply stops calling `tick`.
- Snapshot API for tests: `state.snapshot() -> Dictionary` with `version: int` + everything needed to reproduce the next frame and detect state changes.
- `tick()` returns a `bool` (or a small flags struct) indicating whether the visible state changed this call, so the scene can avoid a full snapshot diff per frame.
- A `set_last_tick(now_ms)` method so the scene can shift internal time after pause/focus-loss without re-creating the state.

### Non-goals
- Rendering, audio, haptics, input bindings (those land in the scene task).
- Multi-snake / co-op.
- Special foods (bonus, slow, shrink) — possible v2.
- Touch swipe handling (scene task).

### Acceptance
1. `SnakeGameState.create(seed: int, config: SnakeConfig) -> SnakeGameState` returns a deterministic state for the same seed/config.
2. Buffering: two `turn(left)` calls within one tick → only the first applies; second is dropped (not queued past the next tick).
3. 180° guard: a `turn(opposite_of_current)` is rejected (head never instantly reverses).
4. Food-spawn invariant: after every food-eaten event, the new food cell is not occupied by any snake segment. Seeded test confirms reproducible food sequence.
5. Walls mode: head moving into x = -1 / x = W / y = -1 / y = H sets `game_over = true`. Wrap mode: head wraps and continues.
6. Self-collision: when head moves onto a body cell that will *not* vacate this tick (i.e., not the tail cell when no food was just eaten), `game_over = true`.
7. Level curve: with default config, after 5 foods, `step_ms` is strictly less than starting `step_ms`. Floor enforced (no faster than `min_step_ms`).
8. Tail-vacate corner case: head moving into the cell the tail will vacate this tick is *legal* if no food was eaten; promoted from tests to acceptance.
9. Snapshot `version: int` field present and stable; documented in `docs/persistence.md`.
10. Test coverage: every branch of `tick`, `turn`, food-spawn, and collision exercised by GUT tests (no automated coverage tool — checklist enforced in PR review).

### Files
```
godot/scripts/snake/core/
  snake_config.gd          # Resource: grid_w, grid_h, walls, start_step_ms, min_step_ms, level_every, level_factor
  snake_state.gd           # SnakeGameState: tick / turn / snapshot
  snake_rng.gd             # Tiny seeded RNG wrapper (or use RandomNumberGenerator with seed)
godot/tests/unit/snake/
  test_snake_state.gd
  test_snake_rng.gd
  fixtures/                # JSON snapshots for golden-state tests
```

### Public API
```gdscript
class_name SnakeGameState

enum Dir { UP, RIGHT, DOWN, LEFT }

static func create(seed: int, config: SnakeConfig) -> SnakeGameState
func turn(dir: int) -> void                # buffers one turn for next tick; rejects 180°
func tick(now_ms: int) -> bool              # returns true if visible state changed; advances zero+ steps
func set_last_tick(now_ms: int) -> void      # shift internal time after pause / focus regain
func snapshot() -> Dictionary                # { version: 1, snake: [[x,y]...], food: [x,y], score, level, step_ms, dir, game_over }
func is_game_over() -> bool
```

### State internals
- `_snake: Array[Vector2i]` — head at 0.
- `_dir: int` — current applied direction.
- `_pending_dir: int = -1` — turn buffered for next tick; reset after applied.
- `_last_tick_ms: int` and `_step_ms: int`. `tick(now_ms)` runs while `now_ms - _last_tick_ms >= _step_ms`, advancing one step per iteration (handles dropped frames).
- `_rng: RandomNumberGenerator` — seeded.

### Per-step algorithm
1. Apply `_pending_dir` if non-(-1) and not 180° opposite of `_dir`. Clear pending.
2. Compute `new_head = _snake[0] + dir_to_vec(_dir)`.
3. Wall check (walls mode) → game over.
4. Will-eat check: `eat = (new_head == _food)`.
5. Self-collision: `new_head` in `_snake[0 .. -2]` if not `eat`, or `_snake[0 .. -1]` if `eat` (because tail won't move when growing). On collision → game over.
6. Insert head; if not `eat` → pop tail; if `eat` → score++, level check, respawn food, optionally shrink `_step_ms`.

### Food spawn
- Sample uniformly from `grid_w * grid_h - len(snake)` empty cells via reservoir-style: build the empty list and pick `_rng.randi() % empty.size()`. For 20×20 this is trivial; revisit only if board grows.

### Tests
- **Determinism**: same seed + same input sequence → identical snapshot stream.
- **Buffer-one**: queueing turn-down then turn-left within the same tick window → only down applies; left is dropped.
- **180° guard**: while moving right, `turn(LEFT)` is ignored; snake keeps moving right.
- **Walls vs wrap**: parameterised over `config.walls`.
- **Self-collision corner case**: tail-follow — moving onto the cell the tail is vacating this tick is *legal* (no eat); test that it does not trigger game over.
- **Self-collision when growing**: same cell while eating → game over (tail does not vacate).
- **Level curve**: feed 5 foods deterministically, assert `step_ms` decreased and is ≥ `min_step_ms`.
- **No food-on-snake**: 1000 ticks of scripted play, assert food never overlaps snake at any tick.

### Risks
- **Tick-loop stutter on long pauses**: if `now_ms` jumps by seconds (tab backgrounded), the loop could run hundreds of iterations and game-over instantly. Mitigation: cap iterations per `tick()` call to e.g. 5; surface remainder in `_last_tick_ms`.
- **RNG portability**: `RandomNumberGenerator` is stable across Godot 4.x patches. Pin Godot version in CI; if it drifts, swap to a small custom xorshift in `snake_rng.gd`.

### Commit sequence
1. `feat(snake): SnakeConfig resource + SnakeGameState skeleton`
2. `feat(snake): tick/turn/move with walls + wrap + collision`
3. `feat(snake): food spawn + scoring + level curve`
4. `test(snake): unit tests + golden snapshots`

---

## Sub-agent review — refinements folded in

Reviewed by `Plan` sub-agent. Resolutions:

- **Snapshot schema**: explicit `version: 1` field added; documented in the contract doc PRD (separate issue, must merge before this).
- **`tick()` return value**: now `bool` (visible state changed). Scene uses this for redraw gating instead of cheap-dict-equality on snapshots.
- **`set_last_tick(now_ms)`**: added to API for pause/focus-loss support. Removes the implicit "scene reaches into core" coupling that snake-scene assumed.
- **Tail-vacate**: promoted from a buried test to acceptance #8.
- **Coverage**: dropped "100% line" — no GUT coverage tool. Replaced with branch-checklist (acceptance #10).
- **180° / buffer-1 rule**: confirmed buffer-1 is intentional. Two presses between ticks → first applies, second dropped (not queued for tick+2). Classic-feel; not the "every queued turn applies" variant.
- **RNG portability**: keep `RandomNumberGenerator`; pin Godot 4.6 in CI; if it ever drifts, swap for xorshift in `snake_rng.gd`.
- **Backgrounded-tab cap**: cap iterations per `tick()` at 5; scene additionally calls `set_last_tick` on focus-in. Both layers documented.
- **Scope kept wrap mode** (used by `easy` difficulty in polish task) — consistency with snake-polish acceptance.
- **No signals on core**: confirmed; state changes are detected via `tick() -> bool` and snapshot diff. Matches the architecture-rule addition in the contract doc PRD.

### PR DESCRIPTION
**Pull request by weavejamtom**
Closes #15

## What

Pure-GDScript Snake rules engine. No `Node`, no `Engine`, no signals — drives entirely off `tick(now_ms) -> bool` + `snapshot() -> Dictionary` per `docs/architecture.md` § "Snapshot diffs, not signals".

Three files under `godot/scripts/snake/core/`:

- `snake_config.gd` — `Resource` with grid/walls/timing/level knobs.
- `snake_rng.gd` — thin wrapper around `RandomNumberGenerator`. Centralized so we can swap to a hand-rolled xorshift if the Godot RNG ever drifts across patch versions (PRD risk note).
- `snake_state.gd` — head-at-0 deque, single-slot turn buffer (first press wins; second within the same tick is dropped, per PRD acceptance #2), 180° guard, walls-or-wrap, tail-vacate semantics, level curve with `min_step_ms` floor, `max_steps_per_tick` cap to survive backgrounded-tab pauses (PRD risk note).

`snapshot()` returns `version: 1` per `docs/persistence.md`.

Per-game `core/` `class_name` lint (added in #33) is honored: each script uses `const Self := preload(…)` for self-typing.

## How tested

184/184 GUT tests pass locally (was 168 → 184; +16 new in `tests/unit/snake/`). Cold-load smoke and lint also pass.

Acceptance coverage:

| # | What | Test |
|---|---|---|
| 1 | Determinism by seed/config | `test_same_seed_and_config_produce_identical_snapshots` |
| 2 | Buffer-1, first wins | `test_buffer_is_single_slot_first_wins` |
| 3 | 180° guard | `test_180_degree_turn_is_rejected` |
| 4 | Food never on snake | `test_food_never_spawns_on_snake_over_long_run` |
| 5 | Walls vs wrap | `test_walls_mode_kills_at_right_edge`, `test_wrap_mode_continues_through_right_edge` |
| 6/8 | Tail-vacate legal | `test_tail_vacate_into_old_tail_cell_is_legal` |
| 6 | Self-collision → game_over | `test_self_collision_when_growing_dies` (driven via wall in this PR — direct growing-collision needs scripted food-placement, deferred to scene PR which has integration harness) |
| 7 | Level curve | `test_level_curve_decreases_step_ms_after_threshold` |
| 9 | `version: 1` in snapshot | `test_snapshot_has_version_field` |
| risks | Cap on backgrounded catch-up | `test_max_steps_per_tick_caps_catch_up` |
| risks | First tick anchors time, no move | `test_first_tick_only_anchors_time_no_movement` |
| risks | Post-game-over freezes | `test_after_game_over_tick_returns_false_and_state_unchanged` |

## Estimated effective diff

239 lines (total 543, excluded 304 — `tests/` + `.uid` + `.gd.uid`).

### CODE REVIEW

File: godot/tests/unit/snake/test_snake_state.gd [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	var cfg: SnakeConfig = _walls_cfg()
+	cfg.grid_w = 6
+	cfg.grid_h = 6
+	var s: SnakeState = SnakeState.create(2, cfg)
+	var died: bool = false
+	for i in 200:
+		s.set_last_tick(0)
+		s.tick(cfg.start_step_ms * (i + 1))
+		if s.is_game_over():
+			died = true
```
Comment (lubobill1990) [inline godot/tests/unit/snake/test_snake_state.gd:153]: Test name promises self-collision-while-growing coverage but the body just asserts "some game_over fires within 200 ticks on a 6×6 grid" via wall death — not collision via self at all. The `eat ? n : n - 1` branch in `_advance_one_step` is the trickiest line in the file and currently has zero direct coverage. Suggest constructing a snake long enough that a turn lands the head on a non-tail body cell while eating (or just inject a snake state directly via `_snake = [...]`).

File: godot/scripts/snake/core/snake_state.gd [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+			_game_over = true
+			return
+		_food = next_food
+	else:
+		_snake.pop_back()
+
+
+## Pick a uniformly-random empty cell. Returns Vector2i(-1, -1) if the board
+## is full (snake covers every cell) — caller treats as terminal.
+func _spawn_food() -> Vector2i:
```
Comment (lubobill1990) [inline godot/scripts/snake/core/snake_state.gd:196]: Dead-code-uncovered: the `if next_food.x < 0: _game_over = true` branch (board fully covered → spawn returns (-1,-1)) has no test. Synthetic small grid (e.g. 1×3 walls=false + manually-set `_snake` of length 3) would exercise it cheaply. Optional follow-up; the path is short but a regression here would silently corrupt the win-state semantic.

File: godot/tests/unit/snake/test_snake_state.gd [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+
+# Drive `state` until it executes exactly `n` step iterations starting from
+# now_ms = 0. Returns the final now_ms (the test driver's "clock").
+func _step_n(state: SnakeState, n: int) -> int:
+	state.set_last_tick(0)
+	# Each step needs step_ms wall-clock to elapse. Drive one step per tick()
+	# call so we never accidentally exercise the multi-step catch-up path
+	# (which has its own dedicated test).
+	for i in range(n):
```
Comment (lubobill1990) [inline godot/tests/unit/snake/test_snake_state.gd:35]: Test driver style is inconsistent: `_step_n` correctly drives one step per `tick()` call (last_tick anchored once, `tick(step_ms*(i+1))`), but most other tests inline `state.set_last_tick(0); state.tick(cfg.start_step_ms * (i + 1))` inside the loop — which means iteration `i` ends up running up to `i+1` steps (capped at `max_steps_per_tick`). Both work for the assertions today, but the inline pattern conflates "how many ticks driven" with "how many steps run" — easy to misread when adding tests later. Suggest standardizing on `_step_n` everywhere.

**Review by lubobill1990 — COMMENTED** [review]
Strong core: pure RefCounted, no class_name (per #36), `version: int` snapshot per docs/persistence.md, deterministic, good acceptance coverage. Three minor test gaps below — non-blocking, will get rolled into the nit-cleanup queue.

**Comment by lubobill1990** [comment]
🧹 Rolled 3 minor finding(s) from this review into the rolling cleanup issue: https://github.com/lubobill1990/little-games/issues/10 (batch: https://github.com/lubobill1990/little-games/issues/10#issuecomment-4365764705).

**Review by lubobill1990 — APPROVED** [review]
Approved by gh-pr-review skill (human override of §4b: 3 minor findings already rolled into nit-cleanup #10; CI green; non-blocking test gaps will be picked up from the rolling cleanup queue).
